cpp/ql/src/semmle/code/cpp/models/implementations/StdMap.qll

```diff
@@ -52,7 +52,52 @@ class StdMapInsert extends TaintFunction {
 }
 
 /**
- * The standard map `swap` functions.
+ * The standard map `emplace` and `emplace_hint` functions.
+ */
+class StdMapEmplace extends TaintFunction {
+  StdMapEmplace() {
+    this.hasQualifiedName("std", ["map", "unordered_map"], ["emplace", "emplace_hint"])
+  }
+
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    // flow from any parameter to qualifier and return value
+    // (here we assume taint flow from any constructor parameter to the constructed object)
```

Comment:
I think only the last argument to `emplace` or `emplace_hint` affects the values in the map. Looking at the constructors for `std::pair`, there's four forms for taint purposes: the nullary constructor defaults both arguments, the unary constructors take a pair `p` and separately initialize`first` from `p.first` and `second` from `p.second`, the binary constructors take two arguments `x` and `y` and separately initialize`first` from `x` and `second` from `y`, and the ternary constructors take three arguments, one trivial struct and two tuples `s` and `t`, and separately initialize `first` by calling a constructor with an unpacked `s` and `second` by calling a constructor with an unpacked `t`.

Keys are more complicated, but we're currently treating them as not propagating taint.

When the return value is a pair, we also need taint from the qualifier object to the return value, since the first element of the return value is a fresh iterator over the qualifier.

Reply:
Ah, I'd missed that you always `emplace` (or `emplace_hint`) a `std::pair`, whereas `try_emplace` takes separate arguments for the key and (zero or more) constructor arguments.  I shall have to make some changes.  And add taint from the qualifier to the return value.

Reply:
I've pushed some commits that address this.  It will indeed get more complicated if/when we do tainted keys.